numpy/core/include/numpy/npy_cpu.h

```diff
@@ -110,10 +110,11 @@
     information about your platform (OS, CPU and compiler)
 #endif
 
-#if (defined(NPY_CPU_X86) || defined(NPY_CPU_AMD64))
-#define NPY_CPU_HAVE_UNALIGNED_ACCESS 1
-#else
+// int*, int64* should be propertly aligned on ARMv7 to avoid bus error
+#if defined(__arm__) && !(defined(__aarch64__) || defined(_M_ARM64))
 #define NPY_CPU_HAVE_UNALIGNED_ACCESS 0
+#else
+#define NPY_CPU_HAVE_UNALIGNED_ACCESS 1
```

Comment:
Thanks. Does this reject SPARC? I am certain sparc does not support unaligned access, I have no clue about the others.

Reply:
As far as I can see, CPUs that support unaligned access is a lot less. so I will use allow-list here, only X86/AMD64/ARMV8 support unaligned access. 

| Archtecture                                      | support of unaligned access                                  | side effects                                                 |
| ------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
| X86                                              | yes                                                          | [small performance loss](https://stackoverflow.com/questions/12491578/whats-the-actual-effect-of-successful-unaligned-accesses-on-x86) |
| AMD64                                            | yes                                                          | [small performance loss](https://stackoverflow.com/questions/45128763/how-can-i-accurately-benchmark-unaligned-access-speed-on-x86-64/45129784#45129784) |
| Sparc                                            | [no](http://www.catb.org/esr/structure-packing/)             | Exception or incorrect result                                |
| old PowerPC(s390)                                | yes                                                          | big performance loss                                         |
| modern PowerPC                                   | no                                                           | Exception or incorrect result                                |
| <=ARMV7                                          | [no](https://news.ycombinator.com/item?id=18855047)          | Exception or incorrect result                                |
| ARMV8                                            | yes                                                          | small performance loss                                       |
| MIPS                                             | no                                                           | Exception or incorrect result                                |
| [IA64](https://en.wikipedia.org/wiki/IA-64)      | [no](https://elixir.bootlin.com/linux/latest/source/arch/ia64/kernel/unaligned.c) | warning                                                      |
| [PA-RISC](https://en.wikipedia.org/wiki/PA-RISC) | [no](https://elixir.bootlin.com/linux/latest/source/arch/parisc/kernel/unaligned.c) | Exception or incorrect result                                |
| ALPHA                                            | no                                                           | Exception or incorrect result                                |

Reply:
I see it more as a compiler policy more than a CPU feature. IMHO `NPY_STRONG_ALIGNMENT` is more robust since the reason behind this #def is to
avoid bus errors. in fact, armv7 supports unaligned memory access limited to the natural alignment of data types which gives you the ability to work around the misaligned access but of course that will affect performance. 

> I am certain sparc does not support unaligned access, 

sorry for misleading, we can just enable it for NPYV archs for now. somethings like that 
```C
#if NPY_CPU_X86 || NPY_CPU_AMD64 || defined(__aarch64__) || defined(__powerpc64__)
    #define NPY_STRONG_ALIGNMENT 0
#endif
#ifndef NPY_STRONG_ALIGNMENT
    #define NPY_STRONG_ALIGNMENT 1
#endif
````

> EDIT: I also seem to remember that some old PowerPCs support unaligned access but it was very slow, so that it was not a good idea to use it?

If you mean 64-bit unaligned access, well yeah but that also the same issue in any
CPUs with 32-bit mode. two-word load, and two-word store.

Reply:
@Qiyu8, didn't see your last comment, could you provide a reference for this table? and how compilers deal with it?

Reply:
This table is made by myself, all of the data is collected from stackoverflow/officer website, I found that the unaligned access information(especially the compiler strategy) is not straightforward , some of the data is deduced from [Linux](https://github.com/torvalds/linux/blob/master/include/asm-generic/unaligned.h), which may be the most general implementation of unaligned access. I basically agree with your macro defined above.  but I read from [IBM ](https://developer.ibm.com/technologies/systems/articles/pa-dalign/) that modern powerpc don't support unaligned access, should we remove `__powerpc64__`?

Reply:
> I found that the unaligned access information(especially the compiler strategy) is not straightforward

because there's a difference between loading memory on bare-metal and under os(userspace). the kernel can handle CPU exception internally without raising errors and also some CPUs allowing control register for trap unaligned access but of course, that will affect performance.

> I basically agree with your macro defined above. but I read from IBM that modern powerpc don't support unaligned access, should we remove __powerpc64__?

64-bit mode of Power supports 64-bit unaligned memory access so you should keep it, also as far as I know all PowerPC chips support 32-bit unaligned memory access and the compiler usually handle 64-bit via two-word routine to avoid the plenty of letting the kernel handle it.

Reply:
You lost me a bit, is the name nicer if it is always used with a `!`, or should it be `STRONG_ALIGNMENT_REQUIRED`?There is also this comment in the current code, which makes things even more confusing for me:
```
/*
 * x86 platform works with unaligned access but the compiler is allowed to
 * assume all data is aligned to its size by the C standard. This means it can
 * vectorize instructions peeling only by the size of the type, if the data is
 * not aligned to this size one ends up with data not correctly aligned for SSE
 * instructions (16 byte).
 * So this flag can only be enabled if autovectorization is disabled.
 */
```
but maybe that is part of the reason for the "STRONG_ALIGNMENT". Anyway, I am happy with any solution you can agree on, the only request I would have is adding a comment above the macro pointing out what the "strong alignment" means and if you can maybe move/duplicate part of the above comment also?

Reply:
The alignment is a complicated topic,  I'll do my best to explain each situation clearly.

Reply:
> You lost me a bit, is the name nicer if it is always used with a !, or should it be STRONG_ALIGNMENT_REQUIRED?

Depend on the context however `ALIGNMENT_REQUIRED` or `STRONG_ALIGNMENT` or `STRONG_ALIGNMENT_REQUIRED` gives the same meanining.

> There is also this comment in the current code, which makes things even more confusing for me:

This is comment is no longer accurate, modern x86 chips don't produce performance regression for unaligned memory access and same goes for aarch64 and Power8/9, and modern compilers will autovectorize it anyway.
`NPY_USE_UNALIGNED_ACCESS` already disabled for both cases, we can later replace the current unrolled loops with universal intrinsics and leave this comment as-is for now.

Reply:
@seiko2plus can you explain the last comment again, I could not quite follow.  The comment says that `x86` supports unaligned access, but the C89 (what about C99) standard allows the compiler to assume *aligned* access (I am not sure it matters whether new compilers make use of that).
Or is "not strong alignment" that alignment is not required even when the compiler uses auto-vectorization?

The thing I would be most scared about most is if the code would work fine with the default compiler flags, but potentially break if someone has the silly idea of compiling NumPy with `-O3` to make it faster (or similar flags).